experimental/concurrent/concurrent.go

```diff
@@ -0,0 +1,92 @@
+package concurrent
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+	"golang.org/x/sync/errgroup"
+)
+
+type splitResponse struct {
+	response backend.DataResponse
+	refID    string
+}
+
+// SingleQuery is a single query to be executed concurrently
+// Index is the index of the query in the request
+// PluginContext is the plugin context
+// Headers are the HTTP headers of the request
+// DataQuery is the query to be executed
+type SingleQuery struct {
+	Index         int
+	PluginContext backend.PluginContext
+	Headers       http.Header
+	DataQuery     backend.DataQuery
+}
+
+// SingleQueryData is the function that plugins need to define to execute a single query
+type SingleQueryData func(ctx context.Context, query SingleQuery) (res backend.DataResponse)
+
+// QueryData executes all queries from a request concurrently, using the provided function to execute each query.
+// The concurrency limit is set by the limit parameter. A negative limit means no limit.
+func QueryData(ctx context.Context, req *backend.QueryDataRequest, fn SingleQueryData, limit int) (*backend.QueryDataResponse, error) {
```

Comment:
Not sure allowing unlimited is wise. Maybe default to amount of cpu's and making the limit optional perhaps?

Reply:
I guess I'm thinking more for any new potential datasource starting to use this, we want the API to guide the user to do good/use a limit which is sane and not unlimited. The limit is for protecting us/grafana against users doing bad things and why we want to limit it. For developer they just want to concurrently run queries and might not think of bad user intentions. Hard limit of 10 sounds like a good idea to begin with.

Note it's not only slow HTTP responses it's the parsing of the responses as well which could be very CPU intensive.